pkg/controller/secret/secret_controller.go

```diff
@@ -0,0 +1,170 @@
+
+// ReconcileSecret reconciles a Secret object
+type ReconcileSecret struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a Secret object and makes changes based on the state read
+// and what is in the Secret.Spec
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileSecret) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling " + userDataSecret)
+
+	userData, err := secrets.GenerateUserData()
+	if err != nil {
+		return reconcile.Result{}, errors.Wrapf(err, "error generating userData secret")
+
+	}
+	// Fetch UserData instance
+	err = r.client.Get(context.TODO(), kubeTypes.NamespacedName{Name: userData.Name, Namespace: userData.Namespace}, &core.Secret{})
+	if err != nil && k8sapierrors.IsNotFound(err) {
+		// Secret is deleted
+		reqLogger.Info("UserData secret not found, recreating the secret.")
+		err = r.client.Create(context.TODO(), userData)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		// Secret created successfully - don't requeue
+		return reconcile.Result{}, nil
+	} else if err != nil {
+		reqLogger.Error(err, "error retrieving the userData secret")
+		return reconcile.Result{}, err
+	} else {
+		// secret is updated
+		reqLogger.Info("Restoring the userData secret")
+		err = r.client.Update(context.TODO(), userData)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		// Secret updated successfully
+		return reconcile.Result{}, nil
+	}
+}
+
+// createValidUserData creates userData secret if it is not present in the required namespace.
+// If the secret exists in the namespace, it validates the secret content and update it as required.
+func (r *ReconcileSecret) createValidUserData() error {
+	// generate valid userData secret
+	validUserDataSecret, err := secrets.GenerateUserData()
+	if err != nil {
+		return errors.Wrapf(err, "error generating valid userData secret")
+	}
+	userData := core.Secret{}
+	err = r.client.Get(context.TODO(), kubeTypes.NamespacedName{Name: userDataSecret,
+		Namespace: userDataNamespace}, &core.Secret{})
+	if err != nil && k8sapierrors.IsNotFound(err) {
+		// Create valid userData secret
+		if err = r.client.Create(context.TODO(), validUserDataSecret); err != nil {
+			return errors.Wrapf(err, "error creating userData secret")
+		}
+	} else if err != nil {
+		return errors.Wrapf(err, "error retrieving userData secret")
+	}
+	// Update userData secret if the content is invalid
+	if string(userData.Data["userData"]) != string(validUserDataSecret.Data["userData"]) {
```

Comment:
Do you need this check now that you checking in the predicate? Or is there some corner case that is not obvious? If there is a corner case that you are accounting for, please mention it here.

Reply:
This func is called in `newReconciler`.  There might be a case where user stops the operator and creates a secret with invalid contents. In that case, we need to check this condition.

Reply:
Won't that cause an update event which is caught by the operator when it comes up?

Reply:
I did not see any Update event when I tried this. Still see [this](https://github.com/openshift/windows-machine-config-operator/pull/92#discussion_r470152208), if the check is not in place

Reply:
I am a bit puzzled by this. Please leave it as it is for now and we can discuss this during parking lot.

Reply:
Discussed about this in the parking lot. This usecase will generate a `Create` event.  I will modify the controller to handle create events.